streams: Show warning when unsubscribing a private stream.

sahil839 opened the pull request:
We show a modal as a warning when unsubscribing a private stream
because it is a reversible action and one cannot re-subscribe to
it until added by other members of the stream.

I have used two different classes for the button in modal because we show
success message below the add subscribers input only when we unsubscribe 
from subscriber list.

<!-- What's this PR for?  (Just a link to an issue is fine.) -->


<!-- How have you tested? -->


**GIFs or Screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
![unsubscribe](https://user-images.githubusercontent.com/35494118/89590389-6c63d580-d865-11ea-979a-ac2e8832ba30.gif)

<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

sahil839 commented:
Yes, I had discussed this bug with tim previously but we did not come to conclusion that what should be the correct behavior. Will remind Tim about this again.

Gittenburg commented:
Two questions regarding keyboard accessibility: Do you focus the confirmation button when you open it? Can you close the dialog with Escape?

sahil839 commented:
Escape button closes the dialog. I am fixing the focus part, Thanks for pointing out.

chdinesh1089 commented on static/js/overlays.js:
```diff
@@ -201,8 +201,6 @@ exports.close_modal = function (selector) {
 
     const elem = $(selector).expectOne();
     elem.modal("hide").attr("aria-hidden", true);
-    // Enable mouse events for the background as the modal closes.
-    exports.enable_background_mouse_events();
```
This is causing models that we have in settings/manage organization like "change password" or "change name" etc to not re-enable mouse events even after the modal is closed. When the puppeteer test changes the password(then waits for the modal to be closed) and then clicks on "alert words" section but the I think the click isn't working because of this change.

(Also tested your PR manually to verify that clicks are actually not re-enabled)

sahil839 commented:
I have fixed the focus part. 

Also, I have added a commit to call `enable_background_events` on triggering of `hidden.bs.modal` and avoid calling it from different places for handling cases of closing it from the close button, etc. This also fixes the bug where mouse events were not enabled when the modal was closed by the Escape button. Had to add a couple of prep commits for this part.

sahil839 commented:
I am not able to figure out why the puppeteer tests are failing. The tests are failing locally too. Locally, sometimes get_api_key test is failing and sometimes change password test is failing, screen shot shows that modal is not opening while in development environment the modal opens on click, which means that the pointer events are enabled. 
@chdinesh1089 do you have idea?

sahil839 commented:
@timabbott Updated the PR, ready for review.
- Added a commit to change the puppeteer tests as a prep commit for enabling pointer-events using `hidden.bs.modal` event. We would not need this if we use `hide.bs.modal` event which would be same as before this PR, as bootstrap docs mentions that after calling `.hide` method on modal, the caller returns before the `hidden.bs.modal` event is executed and before this PR also we were just enabling the events just after calling `.hide` on modal. But logically it feels correct to do so after `hidden.bs.modal` which is executed only after the modal is completely hidden.

andersk commented:
The “Use 'hidden.bs.modal' event” commit can be marked “Fixes: #16688”.

sahil839 commented:
> The “Use 'hidden.bs.modal' event” commit can be marked “Fixes: #16688”.

Done.

sahil839 commented:
@timabbott I have updated the PR. 
One thing that I want to point out here is that one commit removes the code which removes the deactivate stream modal from DOM on hiding (I have written the exact reason for this in commit message) . 
Though we can simply remove the modal element from DOM using `hidden.bs.modal` event, but there is inconsistency in modals behavior. Some modals are always present in DOM without opening them, some are added on opening the element and some are removed after closing the modal. So, I think we can decide a consistent behavior here.

timabbott commented:
@sahil839 can you rebase this?

I think for the modal behavior, strategically I think we should be moving them all to the `config_dialog` module system, so that we can make the decision once and have it apply to all of them.

sahil839 commented:
@timabbott Updated the PR to use confirm_dialog now and is ready for review.
- First commit is just for fixing the behavior as described in commit message and discussed [here](https://chat.zulip.org/#narrow/stream/9-issues/topic/remove.20the.20stream.20row.20on.20unsubscribing).
- Second commit is to focus the confirmation button in modal. We should do this for other modals (focus input for modals which have input and confirmation button for others) also that are not opened using `confirm_dialog`.
- Third commit renames a class such that a common handler can be used for modals. (Explained in detail in commit message)
- Fourth commit also adds a class as a prep commit to the commit that removes the code for removing the modal element from DOM. This also helps in using the same click handler.
- Fifth commit is also for similar case as that of fourth commit.
- SIxth commit is the commit that adds the code for having a modal when unsubscribing private stream. This covers all ways of unsubscribing.
- Seventh commit is to do not remove the modals from DOM after closing such that we can use the `hidden.bs.modal` event. Modals using `confirm_dialog` are stilled remove from DOM but those are using the `hidden.bs.modal` event so it does not cause other problems, but other modals were removed just after `close_modal` or using `hide.bs.modal`, which means at the time when modal was not completely hidden while we enable the mouse events after the modal is completely hidden. (Explained more in commit message)
- Eight commit is to fix the puppeteer tests to wait till mouse events are enabled. This is a prep commit for enabling mouse events using `hidden.bs.modal` event.
- Ninth commit is to enable the background mouse events using `hidden.bs.modal` event.

Some things which need to be done further -
- Showing modal when guest unsubscribes because this is also irreversible. There was some discussion related to this on czo as well.
- Using `confirm_dialog` for some other modals as well.
- Discuss when to remove modals from DOM and when not.

timabbott commented:
> other modals (focus input for modals which have input and confirmation button for others) also that are not opened using confirm_dialog

Probably the right plan is to just migrate similar things to `confirm_dialog`.

> Discuss when to remove modals from DOM and when not.

I think there's structurally no reason to have little confirm modals (etc.) in the DOM that aren't displayed -- the only thing they can do is result in bugs when a selector in them is accessed by other code, or bugs where we end up with 7 copies of a given modal slowly accumulating.

timabbott commented:
This is #9254, tweaking the commit to auto-close that.

timabbott commented:
This is great -- merged this with the following small changes to the text and logic for who this applies to.  Thanks @sahil839!  This closes one of our older open issues, and I think will be appreciated by a lot of users.

```
diff --git a/static/js/subs.js b/static/js/subs.js
index 54a313f4f4..7b5e3d8c38 100644
--- a/static/js/subs.js
+++ b/static/js/subs.js
@@ -959,14 +959,15 @@ export function unsubscribe_from_private_stream(sub, from_stream_popover) {
         parent: modal_parent,
         html_heading: i18n.t("Unsubscribe from __stream_name__", {stream_name: sub.name}),
         html_body,
-        html_yes_button: i18n.t("Yes, unsubscribe from this stream"),
+        html_yes_button: i18n.t("Yes, unsubscribe"),
         on_click: unsubscribe_from_stream,
     });
 }
 
 export function sub_or_unsub(sub, from_stream_popover, stream_row) {
     if (sub.subscribed) {
-        if (sub.invite_only) {
+        // TODO: This next line should allow guests to access web-public streams.
+        if (sub.invite_only || page_params.is_guest) {
             unsubscribe_from_private_stream(sub, from_stream_popover);
             return;
         }
diff --git a/static/templates/unsubscribe_private_stream_modal.hbs b/static/templates/unsubscribe_private_stream_modal.hbs
index d5f8a61910..7c36508142 100644
--- a/static/templates/unsubscribe_private_stream_modal.hbs
+++ b/static/templates/unsubscribe_private_stream_modal.hbs
@@ -1 +1 @@
-<p>{{t "Once you unsubscribe from this stream, you cannot access this stream and you must be added by a member of the stream to join it again." }} <strong>{{t "Are you sure you want to do this?" }}</strong></p>
+<p>{{t "Once you leave a private stream, you will not be able to rejoin." }}</p>
```

I encourage you to look at porting other settings modals to `confirm_dialog` where appropriate, to just to simplify the set of code that we'd need to consider for thinking about "should we remove modals from the DOM" more broadly.
